### ISSUE TITLE
Use current environment for user-defined functions

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/insect#125]
This is the current behavior:
```
>>> a=2
>>> f(x)=a*x  # global environment is closed over here
>>> f(1)
   = 2
>>> a=3  # has no effect, environment was closed over at function definition
>>> f(1)
   = 2
```

However, it would probably be more useful if the current environment would be used at call-time:
```
>>> a=3
>>> f(1)
   = 3
```

**Comment by fattredd** [comment]
I disagree. If you'd like a function to be able to change in the future, use two parameters like this:
~~~
>>> f(x, y) = x * y
>>> f(1, 2)
   = 2
>>> f(1, 3)
   = 3
~~~

Though it can be tedious to type two parameters every time. Perhaps a better solution would be the ability to set function defaults like this: 
~~~
>>> f(x, y=2) = x * y
>>> f(1)
   = 2  
>>> f(1, 3)  
   = 3
~~~

**Comment by sharkdp** [comment]
I initially implemented it like this, because I thought that might be a good idea (and "mutability is bad"). However, most scripting languages (and I tend to think of Insect as a simple scripting language) would imitate the mutable behavior:

Python:
``` python
>>> a = 2
>>> def f(x): return a * x
... 
>>> a = 3
>>> f(1)
3
```

JavaScript:
``` js
> var a = 2;
> function f(x) { return a*x; }
> a = 3;
> f(1)
3
```

It's even possible to define the function `f` before `a` is defined. That's currently not possible in Insect:
```
>>> f(x) = a * x
>>> a = 2
>>> f(1)

  Unknown identifier: a
```

Things get even more weird (in my opinion) if two functions are involved. At function definition, Insect captures the entire environment - including all function definitions:
```
>>> f1(x) = x²
>>> f2(x) = 2 f1(3 x)    # environment (including f1) is stored here
>>> f2(4)
   = 288
>>> f1(x) = x³    # this change does not have any effect
>>> f2(4)
   = 288
```

Coming back to your point: A way to secure a function definition from being changed in the future would be to only use *constants*, not *global variables*. 

(JavaScript):
``` js
> const a = 2;
> function f(x) { return a*x; }
> a = 3;
TypeError: Assignment to constant variable.
```

Currently, Insect only supports built-in constants (like `pi`, `speedOfLight`, etc.). But user-defined constants could be a useful addition.

**Comment by sharkdp** [comment]
@fattredd Any thoughts on this?

**Comment by mhmdanas** [comment]
This change makes sense to me. @sharkdp has your opinion on this issue changed? If not, let's add it in v5.9.0.

**Comment by sharkdp** [comment]
Sounds good. My opinion has not changed.
